src/tests/ftest/io/nvme_io.py

```diff
@@ -24,81 +24,52 @@
 from __future__ import print_function
 
 import os
-import sys
-import json
 import traceback
 import uuid
 import re
 import avocado
+from apricot import TestWithServers
 
-sys.path.append('./util')
-sys.path.append('../util')
-sys.path.append('../../../utils/py')
-sys.path.append('./../../utils/py')
-import server_utils
 import write_host_file
 import ior_utils
-import AgentUtils
 import spdk
 
 from general_utils import DaosTestError
-from daos_api import DaosContext, DaosPool, DaosApiError, DaosLog
+from daos_api import DaosPool, DaosApiError
 
-class NvmeIo(avocado.Test):
+class NvmeIo(TestWithServers):
     """
-    Test Class Description:
-        Test the general Metadata operations and boundary conditions.
+    Run IO over NVMe and verify the pool size.
+    :avocado: recursive
     """
-
     def setUp(self):
         self.pool = None
-        self.hostlist = None
         self.hostfile_clients = None
-        self.hostfile = None
         self.out_queue = None
         self.pool_connect = False
-
-        with open('../../../.build_vars.json') as json_f:
-            build_paths = json.load(json_f)
-
-        self.basepath = os.path.normpath(build_paths['PREFIX']  + "/../")
-        self.server_group = self.params.get("name", '/server_config/',
-                                            'daos_server')
-        self.context = DaosContext(build_paths['PREFIX'] + '/lib/')
-        self.d_log = DaosLog(self.context)
-        self.hostlist = self.params.get("servers", '/run/hosts/*')
-        self.hostfile = write_host_file.write_host_file(self.hostlist,
-                                                        self.workdir)
-        self.hostlist_clients = self.params.get("clients", '/run/hosts/*')
+        self.hostlist_servers = self.params.get("test_machines", '/run/hosts/*')
```

Comment:
This is done already in `TestWithServers.setUp()`:
https://github.com/daos-stack/daos/blob/0562c1f9d8c5b842952883870850ec2b6fe0bc0a/src/tests/ftest/util/apricot/apricot/test.py#L148

Reply:
And you need to do this NVMe setup stuff before `TestWithServers.setUp()` can be done, yes?

I think @phender suggested earlier splitting up the `TestWithServers.setUp()` into multiple methods so that this kind of "in the middle of setUp()" sort of thing can be done.  It's also possible that we could use another class, if this NVMe setup is going to become common, or switch/flag/etc. enabling an optional NVMe setup in `TestWithServers.setUp()`.

But I wouldn't block this patch for that.  As much as it would be nice to not start down a road of repeating setup steps in many tests again.

Reply:
So should I move spdk.nvme_setup in to TestWithServers.setUp()?

Reply:
I don't think you can.  How would a call to `setUp()` know that the hardware that the test is going to run on has NVMe?

Probably the most optimisation we can do at this, point, until we see more use of it is to split the `TestWithServers.setUp()` into multiple methods so that you can call the first one, do your NVMe initialisation and then call the second one (i.e. all instead of calling `setUp()`). And then of course modify `TestWithServers.setUp()` to just call both of the methods, one after the other.

Does that make sense?

Reply:
ok will include one more method to do nvme setup